lib/improver/nbhood/nbhood.py

```diff
@@ -281,84 +514,159 @@ def _sum_and_area_for_edge_cases(cube, cells_x, cells_y, i, j):
-                for i in (edge_columns):
-                    neighbourhood_total[j, i], neighbourhood_area[j, i] = (
-                        _sum_and_area_for_edge_cases(
-                            slice_2d, cells_x, cells_y, i, j))
-            for i in range(n_columns):
-                for j in (edge_rows):
-                    neighbourhood_total[j, i], neighbourhood_area[j, i] = (
-                        _sum_and_area_for_edge_cases(
-                            slice_2d, cells_x, cells_y, i, j))
             with np.errstate(invalid='ignore', divide='ignore'):
                 slice_2d.data = (neighbourhood_total.astype(float) /
                                  neighbourhood_area.astype(float))
-            slice_2d.data[nan_mask] = np.NaN
+            slice_2d.data[nan_mask.astype(bool)] = np.NaN
             cubelist.append(slice_2d)
         return cubelist.merge_cube()
 
     @staticmethod
-    def cumulate_array(cube):
+    def _set_up_cubes_to_be_neighbourhooded(cube):
         """
-        Method to calculate the cumulative sum of an m x n array, by first
-        cumulating along the y direction so that the largest values
-        are in the nth row, and then cumulating along the x direction,
-        so that the largest values are in the mth column. Each grid point
-        will contain the cumulative sum from the origin to that grid point.
+        Set up a cubelist containing either the input cube, or the input cube
+        and a mask cube.
 
         Parameters
         ----------
         cube : Iris.cube.Cube
-            Cube to which the cumulative summing along the y and x direction
-            will be applied.
+            Cube that will be checked for whether the data is masked.
 
         Returns
         -------
-        cube : Iris.cube.Cube
-            Cube to which the cumulative summing along the y and x direction
-            has been applied.
-        nan_mask : numpy array
-            Mask of where the original input data array had nans. Nans will
-            be reapplied at these points after neighbourhood processing has
-             been carried out.
+        cubes_to_sum : Iris.cube.CubeList
+            CubeList containing either the input cube, or the input cube and
+            a mask cube.
         """
-        yname = cube.coord(axis="y").name()
-        xname = cube.coord(axis="x").name()
-        cubelist = iris.cube.CubeList([])
-        for slice_2d in cube.slices([yname, xname]):
-            data = slice_2d.data
-            nan_mask = np.isnan(data)
-            data[nan_mask] = 0
-            data_summed_along_y = np.cumsum(data, axis=0)
-            data_summed_along_x = (
-                np.cumsum(data_summed_along_y, axis=1))
-            slice_2d.data = data_summed_along_x
-            cubelist.append(slice_2d)
-        return cubelist.merge_cube(), nan_mask
+        # If there is no mask, make a mask of ones. This will speed up the
+        # calculation.
```

Comment:
This confuses me a little bit. You check if its masked, and if it is you do something, otherwise you don't. So in the comment where it says 'if there is no mask make a mask of ones', what do you mean?